### PR TITLE
Wifi configuration

### PR DESCRIPTION
**Pull request by danybony**
## Problem

We want to allow the user to automatically configure the conference wifi

## Solution

I used Firebase Remote config to get the feature details (feature enabled/disabled, ssid, password).
If the feature is enabled, we show a new preference in the app settings. To avoid a flickering of that preference in case of feature disabled (we remove it in that case) I choose to fetch the remote config boolean immediately, instead of using `RemoteConfig#getBoolean` (which will fetch updated values and asynchronously emit the actual boolean).

When the user clicks on the preference, we attempt to set the wifi immediately. In case off success, a snackbar with the success message is shown, otherwise we navigate to the wifi config error screen.

**Note**: the layout for this screen needs to be tweaked, as I couldn't match the design using a constraint layout.

In that screen the user can see the network details, copy the password to the clipboard and navigate to the system wifi settings to enter it manually.

#### TODO:
Add analytics

### Test(s) added

No

### Screenshots

 App settings with config success | Error screen | Error screen with password copied message
 ------ | ----- | -----
device-2018-04-09-110558 | image | image

### Paired with

Nobody

### CODE REVIEW

File: app/src/main/java/net/squanchy/remoteconfig/RemoteConfig.kt [hunk 1]
```diff
@@ -14,6 +14,12 @@ class RemoteConfig(
     private val cacheExpiryInSeconds: Long
         get() = if (debugMode) EXPIRY_IMMEDIATELY else EXPIRY_ONE_HOUR
 
+    fun wifiAutoConfigEnabledNow() = firebaseRemoteConfig.getBoolean(KEY_WIFI_CONFIG_ENABLED)
+
+    fun wifiSsid(): String = firebaseRemoteConfig.getString(KEY_WIFI_SSID)
```
Comment (rock3r) [inline app/src/main/java/net/squanchy/remoteconfig/RemoteConfig.kt]: Why not returning a `data class WifiConfiguration(ssid: String, password: String)` for better typing? Would work especially well if you also moved these to a specialised function in the `remoteconfig` package that gets in the `RemoteConfig` instance and avoids having feature-specific constants and methods on `RemoteConfig`, which is supposed to be a feature-agnostic entry point.

```kotlin
internal fun obtainVenueWifiConfiguration(remoteConfig: RemoteConfig): WifiConfiguration? = ...

data class WifiConfiguration(ssid: String, password: String)
```

File: app/src/main/java/net/squanchy/settings/SettingsFragment.kt [hunk 2]
```diff
@@ -89,6 +100,35 @@ class SettingsFragment : PreferenceFragment() {
...
+
+        if (remoteConfig.wifiAutoConfigEnabledNow()) {
+            wifiPreference.setOnPreferenceClickListener { setupWifi(); true }
+        } else {
+            val settingsCategory = findPreference(getString(R.string.settings_category_key)) as PreferenceCategory
+            settingsCategory.removePreference(wifiPreference)
+        }
+    }
+
+    private fun setupWifi() {
```
Comment (rock3r) [inline app/src/main/java/net/squanchy/settings/SettingsFragment.kt:113]: This should really be moved to a collaborator

File: app/src/main/res/values/strings.xml [hunk 3]
```diff
@@ -123,4 +125,13 @@
 
   <string name="schedule_empty_view_text">Nothing here. Come back soon!</string>
 
+  <string name="wifi_config_error_password">Password</string>
+  <string name="wifi_config_error_access_point_name">Access point name</string>
+  <string name="wifi_config_error_title">Oh… uh, it didn\'t work</string>
```
Comment (rock3r) [inline app/src/main/res/values/strings.xml]: You could use a proper apostrophe and avoid the escape = `’` vs `'`

**Review by rock3r — CHANGES_REQUESTED** [review]
Couple of notes, and waiting to get the finalised layout/styles

File: app/src/main/res/values/strings.xml (hunk 3 above)
Comment (danybony) [reply app/src/main/res/values/strings.xml]: https://github.com/squanchy-dev/squanchy-android/pull/529/commits/986c6bf32b6175365425dc462287bb3d5699f5f3

File: app/src/main/java/net/squanchy/remoteconfig/RemoteConfig.kt (hunk 1 above)
Comment (danybony) [reply app/src/main/java/net/squanchy/remoteconfig/RemoteConfig.kt]: https://github.com/squanchy-dev/squanchy-android/pull/529/commits/c29c90ee977defd9cab98be44f2a70761895e019

File: app/src/main/java/net/squanchy/settings/SettingsFragment.kt (hunk 2 above)
Comment (danybony) [reply app/src/main/java/net/squanchy/settings/SettingsFragment.kt:113]: https://github.com/squanchy-dev/squanchy-android/pull/529/commits/149c9fc9646b6a5133e13894202c914bcf0a150b

**Review by rock3r — APPROVED** [review]
Nice one, thanks Danielo! :shipit:

**Pull request merged**